src/cli/self_update/shell.rs

```diff
@@ -0,0 +1,199 @@
+            content: POSIX_ENV,
+        }
+    }
+
+    fn source_string(&self) -> Result<String> {
+        Ok(format!(r#"source "{}/env""#, canonical_cargo_home()?))
+    }
+}
+
+struct Posix;
+impl UnixShell for Posix {
+    fn does_exist(&self) -> bool {
+        true
+    }
+
+    fn rcfiles(&self) -> Vec<PathBuf> {
+        match utils::home_dir() {
+            Some(dir) => vec![dir.join(".profile")],
+            _ => vec![],
+        }
+    }
+
+    fn update_rcs(&self) -> Vec<PathBuf> {
+        // Write to .profile even if it doesn't exist.
+        self.rcfiles()
+    }
+}
+
+struct Bash;
+
+// Bash is the source of many complications because it loads ONLY 1 rc,
+// either a login profile or .bashrc, not both.
+impl Bash {
+    // Bash will load only one of these, the first one that exists, when
+    // a login shell starts. BUT not all shells start inside a login shell!
+    fn profiles() -> Vec<PathBuf> {
+        [".bash_profile", ".bash_login", ".profile"]
+            .iter()
+            .filter_map(|rc| utils::home_dir().map(|dir| dir.join(rc)))
+            .collect()
+    }
+
+    // Bash will load only .bashrc on the start of most GUI terminals.
+    fn rc() -> Option<PathBuf> {
+        utils::home_dir().map(|dir| dir.join(".bashrc"))
+    }
+}
+impl UnixShell for Bash {
+    fn does_exist(&self) -> bool {
+        matches!(process().var("SHELL"), Ok(sh) if sh.contains("bash"))
+            || self.rcfiles().iter().any(|rc| rc.is_file())
+            || matches!(utils::find_cmd(&["bash"]), Some(_))
+    }
+
+    fn rcfiles(&self) -> Vec<PathBuf> {
+        // .bashrc first so it gets skipped in update_rcs
+        let mut profiles = Bash::profiles();
+        if let Some(rc) = Bash::rc() {
+            profiles.push(rc);
+        }
+        profiles
+    }
+
+    fn update_rcs(&self) -> Vec<PathBuf> {
+        Bash::profiles()
+            .into_iter()
+            .filter(|rc| rc.is_file())
+            // bash only reads one "login profile" so pick the one that exists
+            .take(1)
+            // Always pick .bashrc as well for GUI terminals
+            .chain(Bash::rc().filter(|rc| rc.is_file()))
+            .collect()
+    }
+}
+
+struct Zsh;
+impl UnixShell for Zsh {
+    fn does_exist(&self) -> bool {
+        matches!(process().var("SHELL"), Ok(sh) if sh.contains("zsh"))
+            || matches!(process().var("ZDOTDIR"), Ok(dir) if dir.len() > 0)
+            || self.rcfiles().iter().any(|rc| rc.is_file())
+            || matches!(utils::find_cmd(&["zsh"]), Some(_))
+    }
+
+    fn rcfiles(&self) -> Vec<PathBuf> {
+        // FIXME: if zsh exists but is not in the process tree of the shell
+        // on install, $ZDOTDIR may not be loaded and give the wrong result.
+        let zdotdir = match process().var("ZDOTDIR") {
+            Ok(dir) => Some(PathBuf::from(dir)),
+            _ => utils::home_dir(),
+        };
+
+        // Don't bother with .zprofile/.zlogin because .zshrc will always be
```

Comment:
`.zshrc` won't be sourced for the login shell which sets up the environment for things launched from GUI applications, and I believe having a `.zprofile` to set that up is more common than having a `.zshenv`.

Reply:
I know.

GUI terminals can start *without* a login shell, but _do_ source .zshrc, making adding to the login shell irrelevant, and I can point you to the exact lines of code which illustrate that the decision to source .zshrc and .zprofile in zsh are made completely independently of one another:
https://github.com/zsh-users/zsh/blob/19390a1ba8dc983b0a1379058e90cd51ce156815/Src/init.c#1329-1367

.zprofile is excluded from consideration by design.

Reply:
I'm not talking about terminals. I'm talking about logging directly into a DE (which will run under a login shell) and using an application launcher to open an application that ends up calling a non-interactive script. At no point in that chain will there be an interactive shell which loads `.zshrc`, only the one login shell which loads `.zshenv` and/or `.zprofile` and a non-interactive shell which loads only `.zshenv` if it exists. For users without a `.zshenv` this would result in the script running without its path being updated to include `~/.cargo/bin`.

Reply:
I do not believe this hypothetical logic chain actually exists.

Reply:
Nonetheless, I believe this is an excellent argument in favor of resolving this permanently in favor of `.zshenv`, and will update the logic accordingly.

Reply:
I suppose I must concede that this has been an issue before: #556! All the more reason to use zsh's provided solution to this bit of complexity instead of handling it ourselves, in this case.